fuse_viz/include/fuse_viz/mapped_covariance_visual.h

```diff
@@ -0,0 +1,256 @@
+   */
+  Ogre::SceneNode* getOrientationSceneNode()
+  {
+    return orientation_root_node_;
+  }
+
+  /**
+   * \brief Get the shape used to display position covariance
+   * @return the shape used to display position covariance
+   */
+  rviz::Shape* getPositionShape()
+  {
+    return position_shape_;
+  }
+
+  /**
+   * \brief Get the shape used to display orientation covariance in an especific axis
+   * @return the shape used to display orientation covariance in an especific axis
+   */
+  rviz::Shape* getOrientationShape(ShapeIndex index);
+
+  /**
+   * \brief Sets user data on all ogre objects we own
+   */
+  virtual void setUserData(const Ogre::Any& data);
+
+  /**
+   * \brief Sets visibility of this covariance
+   *
+   * Convenience method that sets visibility of both position and orientation parts.
+   */
+  virtual void setVisible(bool visible);
+
+  /**
+   * \brief Sets visibility of the position part of this covariance
+   */
+  virtual void setPositionVisible(bool visible);
+
+  /**
+   * \brief Sets visibility of the orientation part of this covariance
+   */
+  virtual void setOrientationVisible(bool visible);
+
+  /**
+   * \brief Sets position of the frame this covariance is attached
+   */
+  virtual void setPosition(const Ogre::Vector3& position);
+
+  /**
+   * \brief Sets orientation of the frame this covariance is attached
+   */
+  virtual void setOrientation(const Ogre::Quaternion& orientation);
+
+  /**
+   * \brief Sets which frame to attach the covariance of the orientation
+   */
+  virtual void setRotatingFrame(bool use_rotating_frame);
+
+private:
+  void updatePosition(const Eigen::Matrix6d& covariance);
+  void updateOrientation(const Eigen::Matrix6d& covariance, ShapeIndex index);
+  void updateOrientationVisibility();
+
+  Ogre::SceneNode* root_node_;
+  Ogre::SceneNode* fixed_orientation_node_;
+  Ogre::SceneNode* position_scale_node_;
+  Ogre::SceneNode* position_node_;
+
+  Ogre::SceneNode* orientation_root_node_;
+  Ogre::SceneNode* orientation_offset_node_[kNumOriShapes];
+
+  rviz::Shape* position_shape_;                    ///< Ellipse used for the position covariance
+  rviz::Shape* orientation_shape_[kNumOriShapes];  ///< Cylinders used for the orientation covariance
+
+  bool local_rotation_;
+
+  bool pose_2d_;
+
+  bool orientation_visible_;  ///< If the orientation component is visible.
+
+  Ogre::Vector3 current_ori_scale_[kNumOriShapes];
+  float current_ori_scale_factor_;
+
+  const static float max_degrees;
+
+private:
+  // Hide Object methods we don't want to expose
+  // NOTE: Apparently we still need to define them...
+  virtual void setScale(const Ogre::Vector3& scale){};
+  virtual void setColor(float r, float g, float b, float a){};
+  virtual const Ogre::Vector3& getPosition();
+  virtual const Ogre::Quaternion& getOrientation();
+
+  // Make MappedCovarianceProperty friend class so it create MappedCovarianceVisual objects
+  friend class MappedCovarianceProperty;
```

Comment:
Also had to copied the entere covariance visual because it has a private constructor (https://github.com/ros-visualization/rviz/blob/1.12.17/src/rviz/default_plugin/covariance_visual.h#L93) that only the friend class `CovarianceProperty` can call (https://github.com/ros-visualization/rviz/blob/1.12.17/src/rviz/default_plugin/covariance_visual.h#L236). Therefore, even inheriting from `CovarianceVisual` is enough.

This is with `rviz` version 1.12.17. I could try with the latest version from `melodic-devel`, that now has the constructor public: https://github.com/ros-visualization/rviz/blob/melodic-devel/src/rviz/default_plugin/covariance_visual.h#L91
That way, at least I don't have to copy the visual implementation.

Reply:
I found why the latest version of `rviz` had the issue. At least there's a workaround: https://github.com/ros-visualization/rviz/commit/d171d823965e456c6fec2286d6fa10541c84f1bb

Reply:
I've updated this PR with a commit that uses a the simpler `MappedCovarianceVisual` implementation, that requires `rviz` 1.13.4